lib/calculator/parser.rb

```diff
@@ -62,8 +62,63 @@ def shift_tokens
     end
 
     # Entry point for the parser, parses an <em>expression</em>
-    def parse_expression
-      # TODO implement me
-    end
+    	def parse_expression
+		ExpressionNode.new(parse_term,parse_expression_prime)
+	end
+	
+	def parse_expression_prime	
+		case @tokens.first
+			when AddOpToken 
+				ExpressionPrimeNode.new(shift_tokens,parse_term,parse_expression_prime)				
+			when SubtractOpToken
+				ExpressionPrimeNode.new(shift_tokens,parse_term,parse_expression_prime)
+			
+			else 
+				ExpressionPrimeNode.new
+		end
+	end
+
+	def parse_term
+		TermNode.new(parse_factor,parse_term_prime)
+	end
+	
+	def parse_term_prime
+		case @tokens.first
+			when MultiplyOpToken
+				TermPrimeNode.new(shift_tokens,parse_factor,parse_term_prime)
+			when DivideOpToken
+				TermPrimeNode.new(shift_tokens,parse_factor,parse_term_prime)
+			else
+				TermPrimeNode.new
+		end
+	end	
+
+	def parse_factor
+	 	factors = [parse_base]
+  # while loop here which appends elements to factors
+		while @tokens.first.is_a? ExponentOpToken or @tokens.first.is_a? IntegerToken or @tokens.first.is_a? DecimalToken or @tokens.first.is_a? LeftParenthesisToken 
```

Comment:
While this loop seems to work for the tests I have created, I am not sure it's absolutely correct.  I think it's probably best to check for an exponent token and then call both the shift_tokens and parse_factor.  It should be up to the parse_base call to check for integer, decimal, or the left parenthesis.